src/main/scala/vectorpipe/osm/internal/functions/package.scala

```diff
@@ -0,0 +1,676 @@
+      val (sequence, index) = getSequence(i)
+
+      sequence.getX(index)
+    }
+
+    override def getY(i: Int): Double = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getY(index)
+    }
+
+    override def getOrdinate(i: Int, ordinateIndex: Int): Double = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.getOrdinate(index, ordinateIndex)
+    }
+
+    override def size(): Int = _size
+
+    override def setOrdinate(i: Int, ordinateIndex: Int, value: Double): Unit = {
+      val (sequence, index) = getSequence(i)
+
+      sequence.setOrdinate(index, ordinateIndex, value)
+    }
+
+    // TODO this should be invalidated after append (but it doesn't actually matter because all of the appending will
+    // occur ahead of time)
+    private lazy val coordinates: Array[Coordinate] = {
+      val coords = new Array[Coordinate](size())
+
+      for (i <- 0 until size) {
+        coords(i) = getCoordinate(i)
+      }
+
+      coords
+    }
+
+    override def toCoordinateArray: Array[Coordinate] = coordinates
+
+    override def expandEnvelope(env: Envelope): Envelope = {
+      for (i <- 0 until size) {
+        env.expandToInclude(getX(i), getY(i))
+      }
+
+      env
+    }
+
+    override def clone(): AnyRef = {
+      // we're already playing fast and loose
+      this
+    }
+  }
+
+  // join segments together into rings
+  @tailrec
+  private def formRings(
+    segments: GenTraversable[VirtualCoordinateSequence],
+    rings: Seq[CoordinateSequence] = Vector.empty[CoordinateSequence]
+  ): GenTraversable[CoordinateSequence] = {
+    segments match {
+      case Nil =>
+        rings
+      case Seq(h, t @ _ *) if h.getX(0) == h.getX(h.size - 1) && h.getY(0) == h.getY(h.size - 1) =>
+        formRings(t, rings :+ h)
+      case Seq(h, t @ _ *) =>
+        val x = h.getX(h.size - 1)
+        val y = h.getY(h.size - 1)
+
+        formRings(t.find(line => x == line.getX(0) && y == line.getY(0)) match {
+          case Some(next) =>
+            h.append(new PartialCoordinateSequence(next, 1)) +: t.filterNot(line => isEqual(line, next))
+          case None =>
+            t.find(line => x == line.getX(line.size - 1) && y == line.getY(line.size - 1)) match {
+              case Some(next) =>
+                h.append(
+                  new PartialCoordinateSequence(
+                    new ReversedCoordinateSequence(next), 1
+                  )
+                ) +: t.filterNot(line => isEqual(line, next))
+              case None => throw new AssemblyException("Unable to connect segments.")
+            }
+        }, rings)
+    }
+  }
+
+  implicit val geometryFactory: GeometryFactory = new geom.GeometryFactory()
+
+  private def formRings(segments: GenTraversable[LineString])(implicit geometryFactory: GeometryFactory): GenTraversable[Polygon] =
+    formRings(segments.map(_.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
+      .map(geometryFactory.createPolygon)
```

Comment:
Now that this is surfaced by GT, the `implicit val` and arg can go away in favor of accessing it directly.

```scala
  private def formRings(segments: GenTraversable[LineString]): GenTraversable[Polygon] =
    formRings(segments.map(_.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
      .map(GeomFactory.factory.createPolygon)
```

This line was also the impetus for adding some additional `Polygon` constructors (https://github.com/locationtech/geotrellis/pull/2618/files#diff-2268e822bd244d311242d85a54f766d0R40), I think this is actually even simpler and doesn't require access to the `GeometryFactory`:

```scala
  private def formRings(segments: GenTraversable[LineString]): GenTraversable[Polygon] =
    formRings(segments.map(_.getCoordinateSequence).map(s => new VirtualCoordinateSequence(Seq(s))))
      .map(Polygon(_))
```

Reply:
Though, right, these are JTS geometries, not GT ones, so the first is correct.